documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,366 @@
+- Define a Constructor taking a Vector4F<T>
+- A Vector3F<T> Axis property mapping to (X, Y, Z)
+- A T Angle property mapping to 2 * Acos(W)
+- A ref Indexer which takes an int and returns the components in order
+- An AsSpan function which returns this quaternion as a Span of the generic type
+- An IsIdentity property which returns if this Quaternion matches an Identity Quaternion
+- Define `+`, `-`, `*`, and `/` between two Quaternions
+- Define `*` with `T` multiplying each component by the scalar value returning a new quaternion
+- Define unary `~`
+- A Dot function which takes another Quaternion and returns its the dotproduct between them
+  - A static implementation of this function **must** be available
+- A LengthSquared property which returns the dot product of the quaternion with itself
+- A Length property which returns the Square Root of LengthSquared
+- An Invert function inverts the Quaternion
+  - a static Inverse function **must** be available but it returns the inverse rather than affecting the original
+- A Normalize function which normalizes the Quaternion
+  - A static implemenation of this function must be available but returns the normalized Quaternion rather than affecting the original
+- A Concatenate function which takes another Quaternion and concatenates it with this quaternion
+  - A static implementation of this function **must** be available but it returns a new Quaternion rather than affecting the originals
+- A Conjugate function which returns the conjugate of this quaternion
+  - A static implementation of this function **must** be available
+- A static CreateFromAxisAngle function which takes in a Vector3F<T> and an angle and returns a Quaternion representing that rotation
+- A static CreateFromRotationMatrix function which takes either a Matrix3x3 or Matrix4x4 and returns a Quaternion representing that rotation
+- A static CreateFromYawPitchRoll which takes either each components separately or in a Vector3F<T> and outputs a Quaternion representing that rotation
+- A static Lerp function which takes 2 Quaternions and a Scalar matching the generic type which linearly interpolates between the 2 Quaternions with scalar used as the amount to lerp
+- A static SLerp function which takes 2 Quaternions and a Scalar matching the generic type which Spherical linearly interpolates between the 2 Quaternions with scalar used as the amount to lerp
+- A static Zero Quaternion Property
+- A static Identity Quaternion Property
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2F
+- Box2I
+- Box3F
+- Box3I
+- CircleF
+- CircleI
+- Rect3F
+- Rect3I
+- PlaneF
+- PlaneI
+- Ray2F
+- Ray2I
+- Ray3F
+- Ray3I
+- Rect2F
+- Rect2I
+- SphereF
+- SphereI
+
+The Box structs are defined by Min and Max Vectors, while Rectangle and Cube are defined by Origin and Size Vectors. These Types **must** be implicitly castable between each other.
```

Comment:
update for RectN names for rectangle and cube
```suggestion
The Box structs are defined by Min and Max Vectors, while the Rect structs are defined by Origin and Size Vectors. These Types **must** be implicitly castable between each other.
```

Reply:
fixed